File: 242.valid-anagram/242.valid-anagram.py
#
# @lc app=leetcode id=242 lang=python3
#
# [242] Valid Anagram
#

# @lc code=start
# Approach 1: Sort both strings and check whether they are equal or not - time O(nlogn) space O(1) - Optimized space solution
# Approach 1: Make frequency counters for both of strings and check whether they are equal or not - Time O(n), Space O(n) n = s+t
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # Optimized for space complexity. O(nlogn),O(1)
        # return sorted(s) == sorted(t)

        if len(s) != len(t):
            return False

        freqS, freqT = {}, {}
        for i in range(len(s)):
            freqS[s[i]] = 1 + freqS.get(s[i], 0)
            freqT[t[i]] = 1 + freqT.get(t[i], 0)
        # return freqS == freqT (same as below)
        for c in freqS:
            if freqS[c] != freqT.get(c, 0):
                return False
        return True

# @lc code=end

File: 242.valid-anagram/242.valid-anagram.js
/*
 * @lc app=leetcode id=242 lang=javascript
 *
 * [242] Valid Anagram
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
    if (s.length !== t.length) return false;

    let freqS = new Map();
    let freqT = new Map();

    for (let i = 0; i < s.length; i++) {
        freqS.set(s[i], 1 + (freqS.get(s[i]) ? freqS.get(s[i]) : 0));
        freqT.set(t[i], 1 + (freqT.get(t[i]) ? freqT.get(t[i]) : 0));
    }

    for (let [key, value] of freqS) {
        if (value !== freqT.get(key)) return false;
    }

    return true;
};
// @lc code=end

